### PR TITLE
show + in symbol results when some results have been omitted by limiting

### LINKED ISSUE #4579: type:symbol search results incomplete
- **Sourcegraph version:** 3.5.0-rc.1 and sourcegraph.com

#### Steps to reproduce:

1. visit https://sourcegraph.com/search?q=repo:%5Egithub.com/sourcegraph/sourcegraph%24+Repos+type:symbol
2. also visit https://sourcegraph.com/search?q=repo:%5Egithub.com/sourcegraph/sourcegraph%24+ListRepos+type:symbol

See that the first search has 13 results, while the second search has 15 results. The first search should be a superset of the second — how can the result set be smaller?

I don't see any indication that there are more results (e.g. no "show more" or "13+").

EDIT: to be clear, the specific bug here is my last point: there just needs to be an indication that there are more results, just like a normal text search, right?

### PR DESCRIPTION
**Pull request by ijt**
Fixes #4579

Test plan: unit tests

After:
Screen Shot 2019-07-11 at 12 35 19

Before:
Screen Shot 2019-07-11 at 12 54 34

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search_symbols.go [hunk 1]
```diff
@@ -103,12 +103,39 @@ func searchSymbols(ctx context.Context, args *search.Args, limit int) (res []*fi
 		})
 	}
 	err = run.Wait()
+	res2 := limitSymbolResults(res, limit)
+	common.limitHit = symbolCount(res2) < symbolCount(res)
+	return res2, common, err
+}
+
+// limitSymbolResults returns a new version of res containing no more than limit symbol matches,
+// and a boolean telling whether the limit was hit.
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/search_symbols.go]: It doesn't return a boolean

File: cmd/frontend/graphqlbackend/search_symbols_test.go [hunk 2]
```diff
@@ -41,3 +42,110 @@ func TestMakeFileMatchURIFromSymbol(t *testing.T) {
...
+				t.Errorf("res2 = %+v, want %+v", res2, wantRes2)
+			}
+		})
+
+		t.Run("limit 2 => unchanged", func(t *testing.T) {
+			res2 := limitSymbolResults(res, 2)
+			if !reflect.DeepEqual(res2, res) {
+				t.Errorf("res2 = %+v, want %+v", res2, res)
+			}
+		})
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/search_symbols_test.go]: Could you add a test case that tests the "only take subset of symbols in a single `*fileMatchResolver`" behavior in `limitSymbolResults`? I think that's already covered with the `limit 0` cases, but I just gave it a shot locally and I think something like that could make the behavior clearer:

```go
	t.Run("two file matches, multiple symbols per file", func(t *testing.T) {
		res := []*fileMatchResolver{
			{
				symbols: []*searchSymbolResult{
					{symbol: protocol.Symbol{Name: "symbol-name-1"}},
					{symbol: protocol.Symbol{Name: "symbol-name-2"}},
				},
			},
			{
				symbols: []*searchSymbolResult{
					{symbol: protocol.Symbol{Name: "symbol-name-3"}},
					{symbol: protocol.Symbol{Name: "symbol-name-4"}},
				},
			},
		}

		t.Run("symbol count is 4", func(t *testing.T) {
			nsym := symbolCount(res)
			if nsym != 4 {
				t.Errorf("symbolCount(res) = %d, want 2", nsym)
			}
		})

		testCases := []struct {
			name  string
			limit int
			want  []*fileMatchResolver
		}{
			{
				name: "limit 0 => no file matches",
				want: nil,
			},
			{
				name:  "limit 1 => one file match with one symbol",
				limit: 1,
				want: []*fileMatchResolver{
					{
						symbols: []*searchSymbolResult{
							{symbol: protocol.Symbol{Name: "symbol-name-1"}},
						},
					},
				},
			},
			{
				name:  "limit 2 => one file match with all symbols",
				limit: 2,
				want: []*fileMatchResolver{
					{
						symbols: []*searchSymbolResult{
							{symbol: protocol.Symbol{Name: "symbol-name-1"}},
							{symbol: protocol.Symbol{Name: "symbol-name-2"}},
						},
					},
				},
			},
			{
				name:  "limit 3 => two file matches with three symbols",
				limit: 3,
				want: []*fileMatchResolver{
					{
						symbols: []*searchSymbolResult{
							{symbol: protocol.Symbol{Name: "symbol-name-1"}},
							{symbol: protocol.Symbol{Name: "symbol-name-2"}},
						},
					},
					{
						symbols: []*searchSymbolResult{
							{symbol: protocol.Symbol{Name: "symbol-name-3"}},
						},
					},
				},
			},
			{
				name:  "limit 4 => two file matches with all symbols",
				limit: 4,
				want: []*fileMatchResolver{
					{
						symbols: []*searchSymbolResult{
							{symbol: protocol.Symbol{Name: "symbol-name-1"}},
							{symbol: protocol.Symbol{Name: "symbol-name-2"}},
						},
					},
					{
						symbols: []*searchSymbolResult{
							{symbol: protocol.Symbol{Name: "symbol-name-3"}},
							{symbol: protocol.Symbol{Name: "symbol-name-4"}},
						},
					},
				},
			},
		}

		for _, tc := range testCases {
			t.Run(tc.name, func(t *testing.T) {
				res2 := limitSymbolResults(res, tc.limit)
				if len(res2) != len(tc.want) {
					t.Errorf("len(res2)=%d, len(want)=%d", len(res2), len(tc.want))
				}

				if !reflect.DeepEqual(res2, tc.want) {
					t.Error(cmp.Diff(res2, tc.want))
				}
			})
		}
	})
```

Feel free to take it or not :)

File: cmd/frontend/graphqlbackend/search_symbols.go [hunk 3]
```diff
@@ -103,12 +103,39 @@ func searchSymbols(ctx context.Context, args *search.Args, limit int) (res []*fi
...
+func limitSymbolResults(res []*fileMatchResolver, limit int) []*fileMatchResolver {
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/search_symbols.go]: It's probably due to the perf work we've been doing, but do you think it's possible to rewrite this function in a way that doesn't allocate a new slice and instead modifies the existing one in place?

Comment (tsenart) [reply cmd/frontend/graphqlbackend/search_symbols.go]: https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating

File: cmd/frontend/graphqlbackend/search_symbols.go (hunk 1 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/search_symbols.go]: Fixed.

File: cmd/frontend/graphqlbackend/search_symbols_test.go (hunk 2 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/search_symbols_test.go]: Nice. Thank you.

File: cmd/frontend/graphqlbackend/search_symbols.go (hunk 3 above)
Comment (ijt) [reply cmd/frontend/graphqlbackend/search_symbols.go]: Doing it in place broke some tests, so I changed it instead to precompute the capacity for res2 to reduce the allocations.

**Pull request merged**